pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -103,13 +104,105 @@ const themes = {
  *
  * Also it is possible to set up theme by property 'theme'.
  */
-export class Terminal extends React.Component {
-    constructor(props) {
+
+export class TerminalState {
+    terminal: Term;
+    element: HTMLDivElement;
+    channel: cockpit.Channel<Uint8Array>;
+
+    constructor(channel: cockpit.Channel<Uint8Array>) {
+        this.terminal = new Term({
+            cols: 80,
+            rows: 1,
+            cursorBlink: true,
+            fontSize: 16,
+            fontFamily: 'Menlo, Monaco, Consolas, monospace',
+            screenReaderMode: true,
+        });
+        this.terminal.loadAddon(new CanvasAddon());
+        this.element = document.createElement("div");
+        this.channel = channel;
+        this.#connectChannel(channel);
+    }
+
+    #connectChannel(channel: cockpit.Channel<Uint8Array>) {
+        channel.addEventListener('message', (_event, data) => {
+            this.terminal.write(data);
+        });
+
+        this.terminal.onData(data => {
+            if (channel.valid) {
+                /* HACK: Ctrl+Space (and possibly other
+                 * characters) is a disaster: While it is U+00A0
+                 * in unicode, with an UTF-8 representation of
+                 * 0xC2A0, the "visible" string in JS is 0x00
+                 * (with TextEncoder, btoa(), and string
+                 * comparison). The internal representation
+                 * retains half of it, and trying to send it to
+                 * the websocket would result in a single 0xA0,
+                 * which is invalid UTF-8 (and causes the session
+                 * to crash). So intercept and ignore such broken
+                 * chars.  See
+                 * https://github.com/cockpit-project/cockpit/issues/21213 */
+                if (data === '\x00') {
+                    console.log("terminal: ignoring invalid input", data);
+                    return;
+                }
+                channel.send(data as unknown as Uint8Array); // XXX
+            }
+        });
+
+        channel.addEventListener('close', (_event, options) => {
+            const term = this.terminal;
+            term.write('\x1b[31m' + (options.problem || 'disconnected') + '\x1b[m\r\n');
+            term.refresh(term.rows, term.rows);
+        });
+    }
+
+    reset(channel: cockpit.Channel<Uint8Array>) {
+        this.channel.close();
+        this.channel = channel;
+        this.#connectChannel(channel);
+        this.terminal.reset();
+        channel.control({
+            window: {
+                rows: this.terminal.rows,
+                cols: this.terminal.cols
+            }
+        } as cockpit.JsonObject as cockpit.ControlMessage);
+        channel.send(String.fromCharCode(12)); // Send SIGWINCH to show prompt on attaching
+    }
+
+    close() {
+        this.channel.close();
+        this.terminal.dispose();
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-b8771fcf-20250814-123207-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)